web-console/src/lib/components/common/table/DataGridPro.tsx

```diff
@@ -0,0 +1,69 @@
+import { reorderElement, replaceElement } from '$lib/functions/common/array'
+import { fullJoin } from 'array-join'
+
+import { DataGridPro, DataGridProProps, GridColDef, GridValidRowModel } from '@mui/x-data-grid-pro'
+
+export type DataGridColumnViewModel = { field: string; width?: number; flex?: number }[]
+
+export type DataGridProMaxProps<Model extends GridValidRowModel = any> = DataGridProProps<Model> & {
+  columnViewModel?: { field: string; width?: number; flex?: number }[]
+  setColumnViewModel?: (model: NonNullable<DataGridProMaxProps<Model>['columnViewModel']>) => void
+  setColumnVisibilityModel?: DataGridProMaxProps<Model>['onColumnVisibilityModelChange']
+  setFilterModel?: DataGridProMaxProps<Model>['onFilterModelChange']
+}
+
+/**
+ * At this time API for DataGridPro is somewhat inconsistent.
+ * This wrapper is designed to provide a declarative API
+ * and a better separation between Grid data and its presentation.
+ * @param props
+ * @returns
+ */
+function DataGridProMax<Model extends GridValidRowModel>({
```

Comment:
can this be done simpler, e.g., just using the functionality here instead of rolling our own logic? https://mui.com/x/react-data-grid/state/#save-and-restore-the-state

Reply:
In this implementation I used a declarative API of DataGridPro, and you are referring to its imperative State API. I hoped to avoid some complexity of mixing declarative and imperative. But I get what you mean; I will try the State API to see if I can get a nicer solution out of it.